lightning/src/ln/peers/handshake/states.rs

```diff
@@ -1,30 +1,888 @@
+		// to the Conduit for future use.
+		conduit.read(&input[bytes_read..]);
+
+		Ok((
+			None,
+			HandshakeState::Complete(Some((conduit, initiator_pubkey)))
+		))
+	}
+}
+
+// https://github.com/lightningnetwork/lightning-rfc/blob/master/08-transport.md#handshake-state-initialization
+fn handshake_state_initialization(responder_static_public_key: &PublicKey) -> (Sha256, Sha256) {
+	let protocol_name = b"Noise_XK_secp256k1_ChaChaPoly_SHA256";
+	let prologue = b"lightning";
+
+	// 1. h = SHA-256(protocolName)
+	// 2. ck = h
+	let chaining_key = concat_then_sha256!(protocol_name);
+
+	// 3. h = SHA-256(h || prologue)
+	let hash = concat_then_sha256!(chaining_key, prologue);
+
+	// h = SHA-256(h || responderPublicKey)
+	let hash = concat_then_sha256!(hash, responder_static_public_key.serialize());
+
+	(hash, chaining_key)
+}
+
+// Due to the very high similarity of acts 1 and 2, this method is used to process both
+// https://github.com/lightningnetwork/lightning-rfc/blob/master/08-transport.md#act-one (sender)
+// https://github.com/lightningnetwork/lightning-rfc/blob/master/08-transport.md#act-two (sender)
+fn calculate_act_message(local_private_ephemeral_key: &SecretKey, local_public_ephemeral_key: &PublicKey, remote_public_key: &PublicKey, chaining_key: ChainingKey, hash: Sha256, act_out: &mut [u8]) -> (Sha256, SymmetricKey, SymmetricKey) {
+	// 1. e = generateKey() (passed in)
+	// 2. h = SHA-256(h || e.pub.serializeCompressed())
+	let serialized_local_public_key = local_public_ephemeral_key.serialize();
+	let hash = concat_then_sha256!(hash, serialized_local_public_key);
+
+	// 3. ACT1: es = ECDH(e.priv, rs)
+	// 3. ACT2: es = ECDH(e.priv, re)
+	let ecdh = ecdh(local_private_ephemeral_key, &remote_public_key);
+
+	// 4. ACT1: ck, temp_k1 = HKDF(ck, es)
+	// 4. ACT2: ck, temp_k2 = HKDF(ck, ee)
+	let (chaining_key, temporary_key) = hkdf5869rfc::derive(&chaining_key, &ecdh);
+
+	// 5. ACT1: c = encryptWithAD(temp_k1, 0, h, zero)
+	// 5. ACT2: c = encryptWithAD(temp_k2, 0, h, zero)
+	chacha::encrypt(&temporary_key, 0, &hash, &[0; 0], &mut act_out[34..]);
+
+	// 6. h = SHA-256(h || c)
+	let hash = concat_then_sha256!(hash, &act_out[34..]);
+
+	// Send m = 0 || e.pub.serializeCompressed() || c
+	act_out[0] = 0;
+	act_out[1..34].copy_from_slice(&serialized_local_public_key);
+
+	(hash, chaining_key, temporary_key)
 }
 
-pub struct ActOneExpectation {
-	pub(super) hash: HandshakeHash,
-	pub(super) chaining_key: [u8; 32],
+// Due to the very high similarity of acts 1 and 2, this method is used to process both
+// https://github.com/lightningnetwork/lightning-rfc/blob/master/08-transport.md#act-one (receiver)
+// https://github.com/lightningnetwork/lightning-rfc/blob/master/08-transport.md#act-two (receiver)
+fn process_act_message(act_bytes: &[u8], local_private_key: &SecretKey, chaining_key: ChainingKey, hash: Sha256) -> Result<(PublicKey, Sha256, SymmetricKey, SymmetricKey), String> {
```

Comment:
very open to suggestions for clearer naming here btw, as I struggled with that in my original PR, too. `process_act_one_and_two_message` could be an alternative, though I'm frankly not sure if that's perhaps even less clear than the current name

Reply:
I think the name is fine. It makes sense to me and there is enough documentation to help future readers understand.